sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -96,108 +96,145 @@ public ProcessBundleHandler(
+      // If we have created the consumers for this PCollection we can skip it.
+      if (pCollectionIdsToConsumers.containsKey(pCollectionId)) {
+        continue;
+      }
+
+      for (String consumingPTransformId : pCollectionIdsToConsumingPTransforms.get(pCollectionId)) {
+        createRunnerAndConsumersForPTransformRecursively(
+            consumingPTransformId,
+            processBundleDescriptor.getTransformsMap().get(consumingPTransformId),
+            processBundleInstructionId,
+            processBundleDescriptor,
+            pCollectionIdsToConsumingPTransforms,
+            pCollectionIdsToConsumers,
+            addStartFunction,
+            addFinishFunction);
+      }
+    }
+
+    createRunnerForPTransform(
+        pTransformId,
+        pTransform,
+        processBundleInstructionId,
+        processBundleDescriptor.getPcollectionsMap(),
+        pCollectionIdsToConsumers,
+        addStartFunction,
+        addFinishFunction);
+  }
+
+  protected void createRunnerForPTransform(
+      String pTransformId,
+      RunnerApi.PTransform pTransform,
+      Supplier<String> processBundleInstructionId,
+      Map<String, RunnerApi.PCollection> pCollections,
+      Multimap<String, ThrowingConsumer<WindowedValue<?>>> pCollectionIdsToConsumers,
       Consumer<ThrowingRunnable> addStartFunction,
       Consumer<ThrowingRunnable> addFinishFunction) throws IOException {
 
-    BeamFnApi.FunctionSpec functionSpec = primitiveTransform.getFunctionSpec();
 
     // For every output PCollection, create a map from output name to Consumer
-    ImmutableMap.Builder<String, Collection<ThrowingConsumer<WindowedValue<OutputT>>>>
+    ImmutableMap.Builder<String, Collection<ThrowingConsumer<WindowedValue<?>>>>
         outputMapBuilder = ImmutableMap.builder();
-    for (Map.Entry<String, BeamFnApi.PCollection> entry :
-        primitiveTransform.getOutputsMap().entrySet()) {
+    for (Map.Entry<String, String> entry : pTransform.getOutputsMap().entrySet()) {
       outputMapBuilder.put(
           entry.getKey(),
-          consumers.apply(
-              BeamFnApi.Target.newBuilder()
-                  .setPrimitiveTransformReference(primitiveTransform.getId())
-                  .setName(entry.getKey())
-                  .build()));
+          pCollectionIdsToConsumers.get(entry.getValue()));
     }
-    ImmutableMap<String, Collection<ThrowingConsumer<WindowedValue<OutputT>>>> outputMap =
+    ImmutableMap<String, Collection<ThrowingConsumer<WindowedValue<?>>>> outputMap =
         outputMapBuilder.build();
 
+
     // Based upon the function spec, populate the start/finish/consumer information.
-    ThrowingConsumer<WindowedValue<InputT>> consumer;
+    RunnerApi.FunctionSpec functionSpec = pTransform.getSpec();
+    ThrowingConsumer<WindowedValue<?>> consumer;
     switch (functionSpec.getUrn()) {
       default:
         BeamFnApi.Target target;
-        BeamFnApi.Coder coderSpec;
+        RunnerApi.Coder coderSpec;
```

Comment:
Readability is really wacky, since each branch of the switch is a conceptually separate block. The fact that they share a scope is a degeneracy inherited from C. Still fine to merge; I'd rather make progress that nitpick readability here.